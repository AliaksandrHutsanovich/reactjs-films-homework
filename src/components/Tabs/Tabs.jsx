import React, { memo, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Tabs as AntdTabs } from 'antd';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

const { TabPane } = AntdTabs;

const Tabs = ({
  tabsConfig,
  animated,
  defaultActiveKey,
  size,
  centered,
  tabBarExtraContent,
  tabBarGutter,
  tabBarStyle,
  onTabClick,
}) => {
  const tabPanes = useMemo(
    () => tabsConfig.map(
      ({
        key,
        renderItemChildren,
        tabTitle,
        children,
      }) => (
        <TabPane key={key} tab={tabTitle}>{renderItemChildren(children)}</TabPane>
      ),
    ),
    [tabsConfig], // it is possible that we need to have additional optimization
  );

  return (
    <AntdTabs
      animated={animated}
      defaultActiveKey={defaultActiveKey}
      size={size}
      centered={centered}
      tabBarExtraContent={tabBarExtraContent}
      tabBarGutter={tabBarGutter}
      tabBarStyle={tabBarStyle}
      onTabClick={onTabClick}
    >
      {tabPanes}
    </AntdTabs>
  );
};

Tabs.propTypes = {
  tabsConfig: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      renderItemChildren: PropTypes.func,
      tabTitle: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node,
      ]),
    }),
  ).isRequired,
  animated: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      inkBar: PropTypes.bool,
      tabPane: PropTypes.bool,
    }),
  ]),
  defaultActiveKey: PropTypes.string,
  size: PropTypes.oneOf([
    'large',
    'default',
    'small',
  ]),
  centered: PropTypes.bool,
  tabBarExtraContent: PropTypes.node,
  tabBarGutter: PropTypes.number.isRequired,
  tabBarStyle: PropTypes.shape({}),
  onTabClick: PropTypes.func,
};

Tabs.defaultProps = {
  animated: false,
  defaultActiveKey: '0',
  size: 'default',
  centered: false,
  tabBarExtraContent: null,
  tabBarStyle: {},
  onTabClick: null,
};

export default memo(Tabs);
