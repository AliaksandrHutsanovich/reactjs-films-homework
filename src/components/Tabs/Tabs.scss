.tabs > div > div {
  flex: none !important;

  > div > div {
    padding: 0px;
  }
}
