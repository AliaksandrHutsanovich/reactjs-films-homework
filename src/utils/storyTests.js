const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
