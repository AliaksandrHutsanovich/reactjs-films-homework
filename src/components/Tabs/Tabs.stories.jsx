import React from 'react';
import Tabs from './Tabs';

export default {
  component: Tabs,
  title: 'components/Tabs',
};

const renderItem = (content) => (
  <p>
    {content}
  </p>
);

const tabsConfig = [
  {
    key: '1',
    renderItemChildren: renderItem,
    tabTitle: 'Tab 1',
    children: 'This is content of tab 1',
  }, {
    key: '2',
    renderItemChildren: renderItem,
    tabTitle: 'Tab 2',
    children: 'This is content of tab 2',
  }, {
    key: '3',
    renderItemChildren: renderItem,
    tabTitle: 'Tab 3',
    children: 'This is content of tab 3',
  },
];

export const tabsExample = () => (
  <Tabs
    tabsConfig={tabsConfig}
    defaultActiveKey="1"
    tabBarExtraContent={<p>This is extra content</p>}
    tabBarGutter={10}
    tabPosition="top"
  />
);
