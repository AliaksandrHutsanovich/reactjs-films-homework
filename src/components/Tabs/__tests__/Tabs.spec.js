import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../Tabs.stories';

describe('<Tabs />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);
});
