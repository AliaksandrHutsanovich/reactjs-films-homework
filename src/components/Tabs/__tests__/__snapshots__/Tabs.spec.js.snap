// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> tabsExample should be rendered properly 1`] = `
<Tabs
  className="tabs"
  defaultActiveKey="1"
  tabBarExtraContent={
    <p>
      This is extra content
    </p>
  }
  tabBarGutter={10}
>
  <TabPane
    tab="Tab 1"
  >
    <p>
      This is content of tab 1
    </p>
  </TabPane>
  <TabPane
    tab="Tab 2"
  >
    <p>
      This is content of tab 2
    </p>
  </TabPane>
  <TabPane
    tab="Tab 3"
  >
    <p>
      This is content of tab 3
    </p>
  </TabPane>
</Tabs>
`;
